src/askui/tools/askui/askui_controller.py

```diff
@@ -1,40 +1,108 @@
 PC_AND_MODIFIER_KEY = Literal['command', 'alt', 'control', 'shift', 'right_shift', 'backspace', 'delete', 'enter', 'tab', 'escape', 'up', 'down', 'right', 'left', 'home', 'end', 'pageup', 'pagedown', 'f1', 'f2', 'f3', 'f4', 'f5', 'f6', 'f7', 'f8', 'f9', 'f10', 'f11', 'f12', 'space', '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z', 'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z', '!', '"', '#', '$', '%', '&', "'", '(', ')', '*', '+', ',', '-', '.', '/', ':', ';', '<', '=', '>', '?', '@', '[', '\\', ']', '^', '_', '`', '{', '|', '}', '~']
 
 
+HAS_ASKUI_COMPONENT_REGISTRY = os.getenv("ASKUI_COMPONENT_REGISTRY_FILE") is not None
 class AskUiControllerServer():
     def __init__(self) -> None:
         self.process = None
 
     def __find_remote_device_controller(self) -> str:
+        if HAS_ASKUI_COMPONENT_REGISTRY:
+            return self.__find_remote_device_controller_by_component_registry()
+        return self.__find_remote_device_controller_by_legacy_path()
+    
+    def __find_remote_device_controller_by_component_registry(self) -> str:
+        component_registry = load_json_file(os.getenv("ASKUI_COMPONENT_REGISTRY_FILE"))
+
+        if component_registry["DefinitionVersion"] != "1":
+            raise ValueError("ValueError! Invalid AskUIComponentRegistry DefinitionVersion format: ", component_registry["DefinitionVersion"])
+        
+
+        installed_packages = component_registry.get('InstalledPackages')
+        if installed_packages is None:
+            raise ValueError("InstalledPackages not found in the component registry.")
+        
+        remote_device_controller_package_id = "{aed1b543-e856-43ad-b1bc-19365d35c33e}"
+        remote_device_controller = installed_packages.get(remote_device_controller_package_id)
+
+        if remote_device_controller is None:
+            raise ValueError("RemoteDeviceController is not installed in the component registry.")
+        
+        execcutables = remote_device_controller.get('Executables')
+        if execcutables is None:
+            raise ValueError("Executables does not exists.")
+        
+        
+        askui_remote_device_controller_path = execcutables.get('AskUIRemoteDeviceController')
+        if askui_remote_device_controller_path is None:
+            raise ValueError("AskUIRemoteDeviceController executables does not exists.")
+        
+        if not os.path.isfile(askui_remote_device_controller_path):
+            raise FileNotFoundError(f"AskUIRemoteDeviceController executable does not exits under '{askui_remote_device_controller_path}'")
+        
+        return askui_remote_device_controller_path
+        
+    def __find_remote_device_controller_by_legacy_path(self) -> str:
```

Comment:
Just to align us in our coding style: When do you use `__` vs. `_`? I usually go for `_` to keep the class open for extension as I find this hard with the name mangling.

Reply:
`__` it's for me a private method, which should only known within a class. e.g. helper methods which are called from mutliple methods.

Reply:
Okay but why not `_`? Because you want to enforce this by having the name mangling? 

I have bad experience with name mangling (the `__`), e.g., when extending a class and overriding extending methods of the subclass which becomes super ugly which is why I use `_` for private methods, functions etc.

My thoughts kind of align with:  https://stackoverflow.com/a/6930223

What do you think about switching to (single) `_` in the future?

Reply:
Did I done this in the part always wrong? [PEP 8](https://peps.python.org/pep-0008/#naming-conventions) 

1. _single_leading_underscore: weak “internal use” indicator. E.g. from M import * does not import objects whose names start with an underscore.
2. __double_leading_underscore: when naming a class attribute, invokes name mangling (inside class FooBar, __boo becomes _FooBar__boo; see below).


And [it says](https://peps.python.org/pep-0008/#method-names-and-instance-variables): "Use one leading underscore only for non-public methods and instance variables."

So __ is only used for class attributes and _ is only used for instance method and variables. 

I think you're right

Reply:
I adapt the code.

Reply:
Nice, thanks :) 

Yeah, for class methods etc. the name mangling makes sense as you don't want to override (or shadow etc.) them with a subclass but really bind them to the class. Especially since you can call classmethods with `self`